### PR TITLE
Attempt removing use of flexbox in the notebook DOM

### PR DESCRIPTION
**Pull request by SylvainCorlay**
This will also make it possible to make use of `break-inside: avoid` CSS media properties for nbconvert webpdf exports.

Besides, it may also improce rendering performances of notebooks.

### CODE REVIEW

**Review by fcollonval — COMMENTED** [review]
please run benchmark

**Comment by jtpio** [comment]
Thanks @SylvainCorlay for opening the PR.

Some of the failing UI tests look relevant (caret not vertically centered):

image

**Comment by fcollonval** [comment]
This impact the _mobile_ styles. It would be good to add UI tests for that to check we are doing ok.

**Comment by jtpio** [comment]
> This impact the _mobile_ styles. It would be good to add UI tests for that to check we are doing ok.

Yes, it seems there is already an issue with 3.2.4 and the mobile styling: https://github.com/jupyterlab/jupyterlab/issues/11516

**Comment by SylvainCorlay** [comment]
The test failure is due to the fact that I use the "inline-box" display to mimick the horizontal flex box instead of an horizontal table.

**Comment by SylvainCorlay** [comment]
I made the following diagrams of the DOM structure (happy to contribute them to the docs).

Screenshot from 2021-11-22 11-14-00

Currently, horizontal flexboxes are use for
 - inputWrapper
 - outputWrapper
 - inputArea
 - outputArea-child

However, in the nbconvert case, there is not collapser, so the only flexboxes are inputArea and outputArea-child, so we could start there.

**Comment by SylvainCorlay** [comment]
please run benchmark

**Review by SylvainCorlay — COMMENTED** [review]
please run benchmark

**Review by SylvainCorlay — COMMENTED** [review]
please run benchmark

**Review by SylvainCorlay — COMMENTED** [review]
please run benchmark

**Comment by SylvainCorlay** [comment]
Visual regression tests are passing.

**Review by fcollonval — APPROVED** [review]
Thanks @SylvainCorlay

**Comment by jtpio** [comment]
> I made the following diagrams of the DOM structure (happy to contribute them to the docs).

It could be interesting to have that diagram in this section of the docs for example: https://jupyterlab.readthedocs.io/en/stable/extension/notebook.html

Although it might become outdated at some point. In that case we could mention "as of JupyterLab 3.2.x" or "4.0.0" depending on where this lands.

**Comment by gabalafou** [comment]
Hi, hope you don't mind me dropping in on this thread out of the blue. 

I leave the following remarks only in a spirit of spreading knowledge, raising awareness, and asking questions; I am not requesting any changes or making recommendations.

I heard about this PR in today's JupyterLab team meeting, and it raised a red flag in my mind, because as a general rule of thumb, the CSS `display` property affects HTML semantics, which in turn is important for accessibility. However, I did a little digging and luckily, in the specific case of `display: table`, I don't think the general rule of thumb applies (in most cases... there are always caveats in this line of work). More on the topic at Adrian Roselli's blog for anyone who wants the nitty-gritty on [tables, CSS display properties, and ARIA.](https://adrianroselli.com/2018/02/tables-css-display-properties-and-aria.html) 

tl;dr-- display:table is probably not an accessibility issue in this PR but please be careful when applying display properties in a non-standard way. 

That said, I can't help but wonder if display:table is an anti-pattern, though perhaps a desirable and/or necessary one for the `break-inside: avoid` declaration. I also wonder if an @media rule would be an appropriate solution here, though I'm not sure I would prefer it, given that I think it is extremely desirable to reduce markup and style variants as much as possible, but I thought I should at least mention it. 

:)

**Comment by SylvainCorlay** [comment]
> Hi, hope you don't mind me dropping in on this thread out of the blue.

Quite the contrary. Many thanks for your feedback.

> That said, I can't help but wonder if display:table is an anti-pattern, though perhaps a desirable and/or necessary one for the break-inside: avoid declaration. I also wonder if an @media rule would be an appropriate solution here, though I'm not sure I would prefer it, given that I think it is extremely desirable to reduce markup and style variants as much as possible, but I thought I should at least mention it.

I have the exact same thought. There is a companion PR in nbconvert https://github.com/jupyter/nbconvert/pull/1679 that overrides the CSS but I would rather have as little as possible in the `@media` rule. The reason is that I fear it may not be as thoroughly tested as the general case and will be more likely to break. More generally, maintaining two versions will be more difficult.

**Comment by krassowski** [comment]
This PR has introduced two significant regressions:
- https://github.com/jupyterlab/jupyterlab/issues/14564
- https://github.com/jupyterlab/jupyterlab/issues/14625

**Comment by SylvainCorlay** [comment]
I am working on a fix for #14564.
